Fixed shadow mapping bugs

NolaDonato opened the pull request:
- use full resolution of depth buffer by packing color channels
- garbage collect shadow map and depth buffers in scene clear
- fixed math in shaders
- fixed matrix computation

GearVRf-Tests PR #14 has proper tests for this fix

Signed off by: Nola Donato nola.donato@samsung.com

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRConfigurationManager.java:
```diff
@@ -142,6 +142,16 @@ void setDockListenerRequired(boolean value){
      */
     public abstract boolean isHmtConnected();
 
+    /**
```
Methods unused; please remove. In any case only the Oculus backend is supposed to support OVR_Multiview at the moment so this needs to be implemented to return always false for all other backend.

liaxim commented on GVRf/Framework/framework/src/main/jni/objects/mesh_jni.cpp:
```diff
@@ -197,7 +197,7 @@ Java_org_gearvrf_NativeMesh_getTexCoords(JNIEnv * env,
 }
 
-JNIEXPORT jcharArray JNICALL
+
```
Return type got lost during merging I presume. Doesn't compile.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRObjectPicker.java:
```diff
@@ -112,4 +112,21 @@ public void doPick()
+     * @param bv2 second bounding volume to test.
+     * @return true if the boxes intersect, false if not.
+     */
+    protected boolean intersect(GVRSceneObject.BoundingVolume bv1, GVRSceneObject.BoundingVolume bv2)
```
Method unused. Is it needed?

liaxim commented on GVRf/Framework/framework/src/main/jni/objects/light.cpp:
```diff
@@ -200,19 +216,35 @@ void Light::generateFBO() {
+    if (LOG_LIGHT) LOGD("LIGHT: create shadow map depth texture %d", depth_texture_->id());
 }
 
+void Light::deleteDepthTexture()
```
The above is applicable to Light::cleanup also.

liaxim commented on GVRf/Framework/framework/src/main/jni/objects/uniform_block.cpp:
```diff
@@ -0,0 +1,528 @@
+
+namespace gvr {
+
+UniformBlock::UniformBlock(const std::string& descriptor) :
```
Still unused but you seem to do want it merged so I guess we've got to.

liaxim commented on light_jni.cpp:
```diff
@@ -0,0 +1,293 @@
+/* Copyright 2015 Samsung Electronics Co., LTD
```
This file should be deleted.

NolaDonato replied:
It is used by GVRLightBase.java

NolaDonato replied:
I removed this file and rebased

liaxim commented:
gvr-shadows: looks weird to me; without this change it looked off but the shadows were reflecting the light source movement, kind of. The shadows are now stuck and still look off to me, in a different way.

gvr-modelviewer2: lightmap_test.fbx: there are missing and/or wrong textures or something; flickering.

NolaDonato commented:
In gvr-shadows, the shadows were not supposed to change with a directional light when it moves - only when it rotates. The shadows changing was actually a bug.
In the current example, the light source is pointing directly away from the camera which doesn't cast any shadows. I will change it to point more towards the ground.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRConfigurationManager.java:
```diff
@@ -142,6 +142,16 @@ void setDockListenerRequired(boolean value){
+     */
+    private static native boolean nativeUsingMultiview(long ptr);
+
+    public final boolean usingMultiview()
```
Please remove the final keyword. The Daydream backend doesn't compile otherwise.

NolaDonato replied:
I removed it.

liaxim commented on GVRf/Framework/framework/src/main/java/org/gearvrf/GVRTexture.java:
```diff
@@ -215,6 +217,54 @@ void signal() {
+     * @param shaderVariable string with name of shader variable to get the texture coordinates.
+     * @throws UnsupportedOperationException if you change variables more than once.
+     */
+    public void setTexCoords(String vertexTexCoord, String shaderVariable) throws UnsupportedOperationException
```
I think it would be better if these were ctor parameters. If you call this for the first time after the shader has been generated it is not going to have effect, right? Not something that should hold up the pull request.

liaxim commented:
I still see this: ``gvr-modelviewer2: lightmap_test.fbx: there are missing and/or wrong textures or something; flickering.``. Is there a problem with the app? The model definitely looks different with this pull request.

liaxim commented:
I don't know what to think of the updated gvr-shadows. @roshanch Please try the updated gvr-shadows and let us know if you think it is ok. Thanks.

roshanch commented:
Sure.! I will check both the samples.

liaxim commented:
If @roshanch gives thumbs up to the new gvr-shadows and lightmap_test.fbx then we can merge. The demos pull request still needs an update to remove the changes to gvr-modelviewer2.

roshanch commented:
@liaxim I checked gvr-shadows sample and doesn't look correct to me. I will work with @apjagdale on that.

NolaDonato commented:
I rebased and squashed commits. I wrote some new unit tests and notice bugs with the shadows when the light is in particular orientations. Still debugging.

liaxim commented:
Thank you very much for squashing the commits

liaxim commented:
Needs rebase also

liaxim commented:
Should we test and review this again? Thanks.

roshanch commented:
Looks like she is still debugging it.

NolaDonato commented:
No, there is still an outstanding bug I am working on.

liaxim commented:
Can we review and test this now?

NolaDonato commented:
Not yet. After I rebased yesterday all the asset loading tests in VRBenchmark only show one eye. I am trying to figure out if my changes caused this or not ☹

NolaDonato commented:
replaced by PR #909
